Support User: Perform a complete page refresh when activating

jordwest opened the pull request:
This performs a full page reload when a support user token is activated/deactivated to ensure all user data is fully refreshed. Between refreshes, the support token is stored in localstorage using the **store** module. On the following boot of Calypso, the User initialization sequence is intercepted before the original user is loaded, to ensure none of the original user's data is loaded.

Additionally, an error message is now displayed when an error occurs while fetching a token.

No other functionality has been changed; the testing procedure is the same (ref: p4TIVU-2OW-p2)

A few notes for review:
 - The Redux store wasn't an option for storing the support token, as we need to check for the token at the earliest point during **User** module initialization, before the Redux store is initialized.
 - Attempts to use **localforage** also caused issues due to its async API and the synchronous nature of the boot process - see #3556

Fixes #3029, fixes #3067, fixes #3068 as well as various other non-reported issues (ref p4TIVU-2OW-p2)

cc @dllh @gwwar @mtias

dllh commented:
I gave this a test drive, and it's working well for me functionally.

gwwar commented on client/layout/index.jsx:
```diff
@@ -179,7 +178,7 @@ export default connect(
 		return {
 			isLoading,
-			isSupportUser: isSupportUser( state ),
+			isSupportUser: state.support.isSupportUser,
```
Were the selector imports becoming too lengthy?

jordwest replied:
Previously this selector was using the support user and token in the Redux store to determine whether support user was active. Now that we're storing those outside the Redux store, I've opted for just a boolean flag in the Redux state to keep track of whether support user is active or not. We could keep the selector, but it would just be:

```js
const isSupportUser = ( state ) => state.support.isSupportUser
```

gwwar commented on client/support/support-user/index.jsx:
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
+		supportUserRestore: supportUser.rebootNormally,
 		supportUserToggleDialog: flowRight( dispatch, supportUserToggleDialog ),
 	}
 }
```
Missing ;

gwwar commented on client/support/support-user/index.jsx:
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
 		showDialog: state.support.showDialog,
+		errorMessage: state.support.errorMessage,
 	}
 }
```
Missing ;

gwwar commented on client/support/support-user/index.jsx:
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
-		supportUserTokenFetch: flowRight( dispatch, supportUserTokenFetch ),
-		supportUserRestore: flowRight( dispatch, supportUserRestore ),
+		supportUserTokenFetch: supportUser.fetchToken.bind( supportUser ),
+		supportUserRestore: supportUser.rebootNormally,
```
This feels a bit weird to be mapping this to class instance methods.

jordwest replied:
Previously this selector was using the support user and token in the Redux store to determine whether support user was active. Now that we're storing those outside the Redux store, I've opted for just a boolean flag in the Redux state to keep track of whether support user is active or not. We could keep the selector, but it would just be:

```js
const isSupportUser = ( state ) => state.support.isSupportUser
```

gwwar commented on client/lib/user/support-user-interop.js:
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
+		debug( 'Rebooting Calypso with support user' );
+
+		store.set( STORAGE_KEY, { user, token } );
+		window.location.reload();
```
While the features are still in progress, we probably need to clear out the sync-handler (cc @retrofox) and any persisted redux state in addition to the page reload.

`User.prototype.clearStoreIfChanged` in `lib/user` should be updated to handle those localforage keys.

jordwest replied:
Could we reset relevant parts of the redux state tree when a `CURRENT_USER_ID_SET` action is dispatched?

gwwar replied:
Good idea. @aduth can you think of any other cases we might run into besides login/logout/switch user where current user id is updated?

gwwar replied:
@jordwest how about we spin up a separate PR for this one. We have a bit of time before `sync-handler `or `persist-redux` are turned on in prod.

jordwest replied:
Actually, this raises another question. Would we even want to clear out the state tree on `CURRENT_USER_ID_SET`? I believe it happens on each page reload (not just login) - so we don't want to clear out the persisted data if the user ID hasn't changed. Maybe there's room for a `CURRENT_USER_ID_CHANGED` action

jordwest replied:
Tracking this in #3586

retrofox replied:
> While the features are still in progress, we probably need to clear out the sync-handler (cc @retrofox) and any persisted redux state in addition to the page reload.

To clean whole sync-handler data we can use `syncHandler.clearAll()`. I was thinking to keep some `public` data for instance public blogs but maybe it doesn't make sense.

gwwar commented on client/lib/user/support-user-interop.js:
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
 
-		if ( wpcom.setSupportUserToken( getSupportUser( state ), getSupportToken( state ) ) ) {
-			onTokenChange();
+		store.remove( STORAGE_KEY );
```
In a future PR we can work around the sync requirement.

gwwar commented on client/lib/user/support-user-interop.js:
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
  */
-export default function( reduxStore ) {
-	const user = new User();
+class SupportUser {
```
We can maybe leave this for a future PR, but since this is basically a singleton, I think we may be able to work around needing a class. Maybe something like:
```javascript
const reduxStoreReadyPromise = new Promise( ( resolve ) => { 
    setReduxStore = ( store ) => resolve( store ); 
} );
const supportUserIsEnabled = config.isEnabled( 'support-user' );
const export fetchToken( user, password ) {
    if( supportUserIsEnabled ) { 
        return reduxStoreReadyPromise.then(
             //...
        );
    }
}
```

jordwest replied:
moved to :arrow_right:  #3587

gwwar commented:
:+1: Nice work! :sparkles: It's really coming along.

One thing we may want to clear up in this PR or in an immediate follow up is that the currentUser id doesn't change back to the original user id after exiting the mode.

As is, I think it's good to :ship: as an outstanding wip.

jordwest commented:
Thanks for the review @gwwar. I've fixed up the typos and will create issues to track the other remaining bits.

For the time being I've fixed the issue with currentUser id not changing back to the original user by just nuking the store completely (as is done in `User` when it detects a change):
```diff
 	rebootNormally() {
 		debug( 'Rebooting Calypso normally' );

-		store.remove( STORAGE_KEY );
+		store.clear();
 		window.location.reload();
 	}
```

If there are no other objections I'll merge shortly

gwwar commented:
As long as you're awake for the deploy, go for it. :smile:
